File: benchmark/perl/fact.pl
use Time::HiRes qw(time);
use constant BASE => 1000000000;

sub mul_small {
    my ($a, $n) = @_;
    my @result;
    my $carry = 0;
    for (my $i = 0; $i < @$a; $i++) {
        my $prod = $a->[$i] * $n + $carry;
        $result[$i] = $prod % BASE;
        $carry = int($prod / BASE);
    }
    while ($carry > 0) {
        push @result, $carry % BASE;
        $carry = int($carry / BASE);
    }
    return \@result;
}

sub big_to_string {
    my ($a) = @_;
    my $result = "$a->[-1]";
    for (my $i = @$a - 2; $i >= 0; $i--) {
        $result .= sprintf("%09d", $a->[$i]);
    }
    return $result;
}

my $start = time * 1000;
my $result = [1];
for my $i (2..10000) {
    $result = mul_small($result, $i);
}
my $end = time * 1000;
print "Result: " . big_to_string($result) . "\n";
printf "Time: %.0fms\n", $end - $start;

File: benchmark/src/lua/fact.lua
local BASE = 1000000000

local function mul_small(a, n)
    local result = {}
    local carry = 0
    for i = 1, #a do
        local prod = a[i] * n + carry
        result[i] = prod % BASE
        carry = prod // BASE
    end
    while carry > 0 do
        result[#result + 1] = carry % BASE
        carry = carry // BASE
    end
    return result
end

local function big_to_string(a)
    local parts = {tostring(a[#a])}
    for i = #a - 1, 1, -1 do
        parts[#parts + 1] = string.format("%09d", a[i])
    end
    return table.concat(parts)
end

local start = os.clock()
local result = {1}
for i = 2, 10000 do
    result = mul_small(result, i)
end
local finish = os.clock()
print("Result: " .. big_to_string(result))
print(string.format("Time: %.0fms", (finish - start) * 1000))
